datalad/cmd.py

```diff
@@ -185,15 +186,15 @@ def __init__(self,
                  ):
 
         command = cmd
-        self.command = [command] if not isinstance(command, List) else command
-        self.path = path
-        self.output_proc = output_proc
-        self.timeout = timeout
-        self.exception_on_timeout = exception_on_timeout
+        self.command: list = [command] if not isinstance(command, List) else command
+        self.path: Optional[str] = path
+        self.output_proc: Optional[Callable] = output_proc
+        self.timeout: Optional[float] = timeout
+        self.exception_on_timeout: bool = exception_on_timeout
 
         self.stdin_queue = None
         self.stderr_output = b""
-        self.runner = None
+        self.runner: Optional[WitlessRunner] = None
```

Comment:
this delayed lazy initialization becomes an issue. Since we allow for `None`, then type checking complains that we are possibly accessing some attributes from `None`:

```
$    mypy --follow-imports skip datalad/cmd.py datalad/runner/runner.py                          
datalad/cmd.py:281: error: "None" has no attribute "runner"
datalad/cmd.py:330: error: "None" has no attribute "return_code"
datalad/cmd.py:355: error: "None" has no attribute "put"
datalad/cmd.py:402: error: "None" has no attribute "send"
datalad/cmd.py:501: error: "None" has no attribute "runner"
Found 5 errors in 1 file (checked 2 source files)
```

how to solve that properly @jwodder @christian-monch ?

Reply:
Unless you want to completely restructure the code, the only real option is to add `assert self.runner is not None` once per method before any uses of `self.runner` where you, the human, can guarantee that `self.runner` has already been filled in.

Reply:
There are two issues here:

1. Type checking fails
2. Initialisation

For the type checker we could just add pure declarations, e.g.:
```
from datalad.runner.nonasyncrunner import _ResultGenerator


    ...


        self.stdin_queue: Queue
        self.generator: _ResultGenerator
``` 
That would satisfy `mypy`.

Since the "declaration" does not create the attributes, we do not even need an `assert self.runner is not None`, because we would get an `AttributeError` if the attributes were accessed before they are initialized.

Reply:
Will push a commit

Reply:
Since I have no write access to your repo, I added PR #7006 which contains your changes and some changes from me on top. PR #7006, should obsolete this PR

Reply:
> Since I have no write access to your repo, 

you should have since I have marked "Allow edits and access to secrets by maintainers" on the right!  Did you try?

> I added PR #7006 which contains your changes and some changes from me on top. PR #7006, should obsolete this PR

works as well for me, thank you!